allel/io/vcfreader.py

```diff
@@ -0,0 +1,255 @@
+import os
+import struct
+import gzip
+import zlib
+import re
+import numpy as np
+
+# max value of int 32
+MAX_INT32 = 0xffffffff >> 1
+
+# max value of uint64
+MAX_UINT64 = 0xffffffffffffffff
+
+
+class VCFReader:
+    def __init__(self, filename, index, region):
+        self._chrom, self._begin, self._end = self.parse_region(region)
+        self._chunk_begin = 0
+        self._chunk_end = MAX_UINT64
+        self._meta = b'#'
+        self._header = True  # track whether we should read VCF header
+        self._block = None
+        self._parse_index(index)
+        self._fileobj = open(filename, 'rb')
+        self._load_block(0)  # load header
+
+    @staticmethod
+    def parse_region(region):
+        """
+        parse a region to chromosome, begin, end
+
+        Parameters
+        ----------
+        region: string
+            tabix style region string
+
+        Returns
+        -------
+        chromosome: bytes
+            contig name
+        begin: int
+            position begin
+        end: int
+            position end
+
+        """
+        try:
+            chrom, begin, end = re.findall(r'^(\w+):*(\d*)-*(\d*)$', region).pop(0)
+        except IndexError:
+            raise ValueError('bad region string: {}'.format(region))
+        begin = int(begin) - 1 if begin else 0
+        end = int(end) if end else MAX_INT32
+        if begin > 0 and end == MAX_INT32:
+            end = begin + 1
+        return chrom.encode('ascii'), begin, end
+
+    @staticmethod
+    def reg2bins(begin, end, n_lvls=5, min_shift=14):
+        """
+        generate key of bins which may overlap the given region,
+        check out https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3042176/
+        and https://samtools.github.io/hts-specs/tabix.pdf
+        for more information.
+
+        Parameters
+        ----------
+        begin: int
+            chromosome position begin
+        end: int
+            chromosome position end
+        n_lvls: int, optional
+            cluster level, for tabix, set to 5
+        min_shift: int, optional
+            minimum shift, for tabix, set to 14
+
+        Returns
+        -------
+        generator
+
+        """
+        begin, end = begin, end
```

Comment:
This line doesn't seem to do anything?

Reply:
My fault, the origin purpose is to convert to zero base region, but I do that in other place but forget to delete this, I will fix that.